packages/asset-swapper/src/utils/market_operation_utils/pools_cache/pair_swaps_cache.ts

```diff
@@ -0,0 +1,91 @@
+import { BalancerSwaps } from '../types';
+
+import { ONE_HOUR_IN_SECONDS, ONE_SECOND_MS } from '../constants';
+
+export interface CacheValue {
+    expiresAt: number;
+    balancerSwaps: BalancerSwaps;
+}
+
+// tslint:disable:custom-no-magic-numbers
+// Cache results for 30mins
+const DEFAULT_CACHE_TIME_MS = (ONE_HOUR_IN_SECONDS / 2) * ONE_SECOND_MS;
+const DEFAULT_TIMEOUT_MS = ONE_SECOND_MS;
+export const EMPTY_BALANCER_SWAPS = { swapInfoExactIn: [], swapInfoExactOut: [] };
+// tslint:enable:custom-no-magic-numbers
+
+/**
+ * Caches SwapInfo for a pair of tokens.
+ * SwapInfo includes swap steps and asset information for those swap steps.
+ */
+export abstract class SwapInfoCache {
+    protected static _isExpired(value: CacheValue): boolean {
+        return Date.now() >= value.expiresAt;
+    }
+    constructor(
+        protected readonly _cache: { [key: string]: CacheValue },
+        protected readonly _cacheTimeMs: number = DEFAULT_CACHE_TIME_MS,
+    ) {}
+
+    public async getFreshPoolsForPairAsync(
+        takerToken: string,
+        makerToken: string,
+        timeoutMs: number = DEFAULT_TIMEOUT_MS,
+    ): Promise<BalancerSwaps> {
+        const timeout = new Promise<BalancerSwaps>(resolve => setTimeout(resolve, timeoutMs, []));
+        return Promise.race([this._getAndSaveFreshSwapInfoForPairAsync(takerToken, makerToken), timeout]);
+    }
+
+    public getCachedSwapInfoForPair(
+        takerToken: string,
+        makerToken: string,
+        ignoreExpired: boolean = true,
+    ): BalancerSwaps | undefined {
+        const key = JSON.stringify([takerToken, makerToken]);
```

Comment:
should we be lowercasing the token addresses in this class? I don't remember whether we do it upstream somewhere

Reply:
I think tokens come in lowercased and, so long as we don't mess up token constants, intermediate tokens should be too.